src/Compilers/Shared/NamedPipeUtil.cs

```diff
@@ -0,0 +1,162 @@
+
+                (string name, bool admin) clientIdentity = default;
+                pipeStream.RunAsClient(() => { clientIdentity = getIdentity(impersonating: true); });
+
+                return
+                    StringComparer.OrdinalIgnoreCase.Equals(serverIdentity.name, clientIdentity.name) &&
+                    serverIdentity.admin == clientIdentity.admin;
+
+                (string name, bool admin) getIdentity(bool impersonating)
+                {
+                    var currentIdentity = WindowsIdentity.GetCurrent(impersonating);
+                    var currentPrincipal = new WindowsPrincipal(currentIdentity);
+                    var elevatedToAdmin = currentPrincipal.IsInRole(WindowsBuiltInRole.Administrator);
+                    return (currentIdentity.Name, elevatedToAdmin);
+                }
+            }
+
+            return true;
+        }
+
+#if NET472
+
+        /// <summary>
+        /// Mono supports CurrentUserOnly even though it's not exposed on the reference assemblies for net472. This 
+        /// must be used because ACL security does not work.
+        /// </summary>
+        private static PipeOptions CurrentUserOption = PlatformInformation.IsRunningOnMono
+            ? (PipeOptions)s_currentUserOnlyValue
+            : PipeOptions.None;
+
+        internal static NamedPipeServerStream CreateServer(
+            string pipeName,
+            PipeDirection direction,
+            int maxNumberOfServerInstances,
+            PipeTransmissionMode transmissionMode,
+            PipeOptions options,
+            int inBufferSize,
+            int outBufferSize) =>
+            new NamedPipeServerStream(
+                pipeName,
+                direction,
+                maxNumberOfServerInstances,
+                transmissionMode,
+                options | CurrentUserOption,
+                inBufferSize,
+                outBufferSize,
+                CreatePipeSecurity(),
+                HandleInheritability.None);
+
+        /// <summary>
+        /// Check to ensure that the named pipe server we connected to is owned by the same
+        /// user.
+        /// </summary>
+        internal static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStream)
+        {
+            if (PlatformInformation.IsWindows)
+            {
+                var currentIdentity = WindowsIdentity.GetCurrent();
+                var currentOwner = currentIdentity.Owner;
+                var remotePipeSecurity = pipeStream.GetAccessControl();
+                var remoteOwner = remotePipeSecurity.GetOwner(typeof(SecurityIdentifier));
+                return currentOwner.Equals(remoteOwner);
+            }
+
+            // Client side validation isn't supported on Unix. The model relies on the server side
+            // security here.
+            return true;
+        }
+
+        internal static PipeSecurity CreatePipeSecurity()
+        {
+            if (PlatformInformation.IsRunningOnMono)
```

Comment:
It looks like mono team suggested this change but is there a way to check more directly "is pipe security supported?"

Reply:
You can see whether it throws. I'm not aware of any feature flags or anything, we throw when you attempt to use those APIs since they're very windows-y, and it doesn't seem like we have any plans to support them.